ishop/views.py

```diff
@@ -1,3 +1,179 @@
-from django.shortcuts import render
+from django.contrib.auth import get_user_model, login
+from django.contrib.auth.mixins import LoginRequiredMixin, UserPassesTestMixin
+from django.contrib.auth.views import LoginView, LogoutView
+from django.shortcuts import redirect
+from django.utils import timezone
+from django.views.generic import ListView, CreateView, UpdateView
+from django.contrib import messages
 
-# Create your views here.
+from ishop.forms import CustomUserCreationForm
+from ishop.models import Good, ShopUser, Purchase, Refund
+
+
+class SuperUserRequiredMixin(LoginRequiredMixin, UserPassesTestMixin):
+    def test_func(self):
+        return self.request.user.is_superuser
+
+
+class GoodsListView(ListView):
+    paginate_by = 10
+    http_method_names = ['post', 'get']
+    template_name = 'goods_list.html'
+    queryset = Good.objects.filter(in_stock__gt=0)
+    extra_context = {'title': 'Online shop'}
+
+    def post(self, request, *args, **kwargs):
+        if not request.user.is_authenticated:
+            msg = "Only logged users can buy"
+            messages.warning(self.request, msg)
+            return redirect('login')
+
+        user = ShopUser.objects.get(id=request.user.id)
+        good = Good.objects.get(id=self.request.POST['pk'])
+        quantity = int(self.request.POST['quantity'])
+        price = int(self.request.POST['price'])
+        in_stock = good.in_stock
+        wallet = user.wallet
+        amount = quantity * price
+
+        if wallet < amount:
+            msg = "You don't have enough money for this purchase"
+            messages.error(self.request, msg)
+            return redirect('goods')
+
+        if in_stock < quantity:
+            msg = "We don't have enough goods in stock"
+            messages.warning(self.request, msg)
+            return redirect('goods')
+
+        user.wallet -= amount
+        user.save()
+        good.in_stock -= quantity
```

Comment:
This action should be atomic

Reply:
totally agree, but I don't know the way
I will ask after the lesson how to organize it

Reply:
done with atomic decorator